public/app/features/logs/logsModel.ts

```diff
@@ -611,15 +611,31 @@ type LogsVolumeQueryOptions<T extends DataQuery> = {
   range: TimeRange;
 };
 
+function defaultExtractLevel(dataFrame: DataFrame): LogLevel {
```

Comment:
This logic may not be applicable to all data sources. I wonder if we may need to be more flexible to allow overriding it as well. At least we should support data plane spec (which I think should read level from `severity`), this way data sources could ensure the level is extracted correctly with proper data frame format.

Reply:
Maybe log volume response should also be part of data plane? 🤔

Reply:
It's applicable to the 2 data sources that currently support logs volume (Loki and Elasticsearch), and can also be overriden by `options`: https://github.com/grafana/grafana/pull/80281/files#diff-cda90ed4a0399053b7808392e0fb6b8f03ed64289acecb0978dc0ba544dd40f8R637

Reply:
Hm... but to override `options` the plugin would need to call `queryLogsVolume` from `logsModel` which wouldn't have access to, right? So there won't be an easy way to override options in plugins code 🤔

Reply:
I think the key here is _may not_. So far we have 2 implementations using logs volume, and both share the same level extraction logic. For now I would say YAGNI. What do you think?

Reply:
I'm just concerned about backwards compatibility, and a dreadful maintenance of multiple returned results if we add one more on top of it when we need it... Do you need getProvider() API to be public or we will keep it internal? I'm also now wondering if we need to support handling Observable returned from `getProvider` as it seems Elastic and Loki will always return a request now 🤔  In theory we could rename `getProvider` to `mapToSupplementaryRequest` 🤔

Reply:
> In theory we could rename getProvider to mapToSupplementaryRequest 🤔

+1 absolutely.

Reply:
I think this logic will actually be applicable for all data sources, that return correct data frames in  `Time Series Wide Format` or `Time Series Multi Format` with `level` label. Maybe instead of `extractLevel` function we just need the way to override `level` label, if it does not have name `level`. That way data sources can override getting of level, and we support it correctly for every data source that produces valid data plane compatible frames. 

And considering log volume hasn't worked for any data source besides Loki and Elasticsearch, we don't need to worry about backward compatibility.

Reply:
> I'm also now wondering if we need to support handling Observable returned from getProvider as it seems Elastic and Loki will always return a request now 🤔

I think this was actually removed and we only expect `DataQueryRequest<TQuery>` and not observable.

```
getDataProvider(
    type: SupplementaryQueryType,
    request: DataQueryRequest<TQuery>
  ): DataQueryRequest<TQuery> | undefined;
  ```

Reply:
`getSupplementaryQueryRequest` sounds great.

Reply:
Apart from Loki and Elasticsearch there's one more internal implementation in [ClickHouse Datasource](https://github.com/grafana/clickhouse-datasource/blob/main/src/data/CHDatasource.ts#L63). 

We need to keep the old method, though it can be marked as deprecated. We ccan introduce a new called `getSupplementaryQueryRequest`. ClickHouse should be updated and only then the old method can be removed, but a big question is how many versions back we need to support 🤔  What are the latest guidelines for internal plugins? Do you own the ClickHouse plugin as well? Would it be possible to use the new `getSupplementaryQueryRequest` and the current level extraction logic in that plugin? 🤔

Reply:
Maybe we could send a PR to update ClickHouse with these changes? I think doing in two steps would be the longer path just for one data source. I wouldn't mind at all doing that as a follow up to the API changes here. What do you all think? Nice catch btw, we almost miss it!

Reply:
It depends on when new version of a plugin is released and how many previous versions of Grafana we need to support. If we release the change in the API before plugin is updated it will stop working. Also, I think plugins are not updated automatically, so I guess there's need to be some grace period with backwards compatibility anyway. What guidelines do you have for newly externalized plugins?